### PR TITLE
Export query result records as JSON

### PR DESCRIPTION
**Pull request by huboneo**
This PR enables users to export result records as JSON, trying to mimic APOC output as much as possible.

### STR
- Run cypher query (eg. `MATCH p = (:Person)-[r]-(:Movie) RETURN p`
- Select "Export JSON" from dropdown

### Screenshots
Screenshot 2019-08-15 at 14 12 05
Screenshot 2019-08-15 at 14 12 15

### Sample output
See unit tests

### CODE REVIEW

File: src/browser/modules/Stream/FrameTitlebar.jsx [hunk 1]
```diff
@@ -60,11 +62,14 @@ import {
...
 import {
   stringifyResultArray,
-  transformResultRecordsToResultArray
+  transformResultRecordsToResultArray,
+  recordToJSONMapper
 } from 'browser/modules/Stream/CypherFrame/helpers'
 import { csvFormat } from 'services/bolt/cypherTypesFormatting'
 import arrayHasItems from 'shared/utils/array-has-items'
 
+const JSON_EXPORT_INDENT = 2
```
Comment (huboneo) [inline src/browser/modules/Stream/FrameTitlebar.jsx]: @oskarhane any suggestions for where to put this?

File: src/browser/modules/Stream/styled.jsx [hunk 2]
```diff
@@ -346,8 +346,9 @@ export const DropdownContent = styled.li`
   box-shadow: 0px 8px 16px 0px rgba(0, 0, 0, 0.2);
   border-radius: 6px;
   z-index: 1;
+  text-align: left;
   line-height: 30px;
-  padding: 5px 0 5px 0;
+  padding: 5px 10px;
```
Comment (huboneo) [inline src/browser/modules/Stream/styled.jsx]: Had to do this to fix dropdown

File: src/browser/modules/Stream/CypherFrame/helpers.js [hunk 3]
```diff
@@ -287,3 +298,86 @@ const arrayifyPath = (types = neo4j.types, path) => {
...
+function neo4jValueToPlainValue (value) {
+  switch (get(value, 'constructor')) {
+    case neo4j.types.Date:
+    case neo4j.types.DateTime:
+    case neo4j.types.Duration:
+    case neo4j.types.LocalDateTime:
+    case neo4j.types.LocalTime:
+    case neo4j.types.Time:
+    case neo4j.types.Point:
+    case neo4j.types.Integer: // not exposed in typings but still there
```
Comment (huboneo) [inline src/browser/modules/Stream/CypherFrame/helpers.js]: Could be that I should call `toInt()` for Integers?

File: src/browser/modules/Stream/CypherFrame/helpers.js [hunk 4]
```diff
@@ -287,3 +298,119 @@ const arrayifyPath = (types = neo4j.types, path) => {
...
+      return value.inSafeRange() ? value.toInt() : value.toNumber()
+    case neo4j.types.Point:
+      return {
+        type: neo4j.types.Point.name,
+        coordinates:
+          value.z !== undefined
+            ? [value.x, value.y, value.z]
+            : [value.x, value.y]
+      }
```
Comment (huboneo) [inline src/browser/modules/Stream/CypherFrame/helpers.js]: Having asked around my suggestion is that we export `Point` as GeoJSON?

File: src/browser/modules/Stream/styled.jsx (hunk 2 above)
Comment (huboneo) [reply src/browser/modules/Stream/styled.jsx]: moved this to DropdownItem to fix appearance

File: src/browser/modules/Stream/FrameTitlebar.jsx (hunk 1 above)
Comment (oskarhane) [reply src/browser/modules/Stream/FrameTitlebar.jsx]: No need to extract this at all imo.

File: src/browser/modules/Stream/CypherFrame/helpers.test.js [hunk 5]
```diff
@@ -726,4 +727,510 @@ describe('helpers', () => {
       ])
     })
   })
+
+  describe('recordToJSONMapper', () => {
```
Comment (oskarhane) [inline src/browser/modules/Stream/CypherFrame/helpers.test.js]: Well done on the tests 🎉

File: src/browser/modules/Stream/CypherFrame/helpers.js [hunk 6]
```diff
@@ -287,3 +298,110 @@ const arrayifyPath = (types = neo4j.types, path) => {
...
+  if (Array.isArray(values)) {
+    return map(values, mapNeo4jValuesToPlainValues)
+  }
+
+  if (isNeo4jValue(values)) {
+    return neo4jValueToPlainValue(values)
+  }
+
+  // could be a Node or Relationship
+  const type = lowerCase(get(values, 'constructor.name', ''))
```
Comment (oskarhane) [inline src/browser/modules/Stream/CypherFrame/helpers.js]: From discussion: call this `elementType`

File: src/browser/modules/Stream/CypherFrame/helpers.js [hunk 7]
```diff
@@ -287,3 +298,110 @@ const arrayifyPath = (types = neo4j.types, path) => {
...
+
+  if (includes(['relationship', 'node'], type)) {
+    const labels =
```
Comment (oskarhane) [inline src/browser/modules/Stream/CypherFrame/helpers.js]: From discussion: Call this `type` from relationships and keep `labels` for nodes.

**Comment by huboneo** [comment]
@oskarhane updated the PR as per your comments

**Pull request merged**